transforms/no-vars.js

```diff
@@ -20,26 +20,87 @@ module.exports = function(file, api) {
+   * contains mutations. Takes an optional VariableDeclarator node argument to
+   * return only whether that specific Identifier is mutated
+   *
+   * @param {ASTPath} node VariableDeclaration path
+   * @param {ASTNode} [declarator] VariableDeclarator node
+   * @return {Boolean}
+   */
+  const isMutated = (node, declarator) => {
+    const scopeNode = node.parent;
+
+    const hasAssignmentMutation = j(scopeNode)
       .find(j.AssignmentExpression)
       .filter(n => {
+        if (declarator) {
+          if (declarator.id.type === 'ObjectPattern') {
+            return declarator.id.properties.some(d => d.value.name === n.value.left.name);
+          } else if (declarator.id.type === 'ArrayPattern') {
+            return declarator.id.elements.some(d =>
+              (d.type === 'RestElement' ? d.argument.name : d.name) === n.value.left.name
+            );
+          }
+
+          if (n.value.left.type === 'ObjectPattern') {
+            return n.value.left.properties.some(p => p.key.name === declarator.id.name);
+          } else if (n.value.left.type === 'ArrayPattern') {
+            return n.value.left.elements.some(e =>
+              (e.type === 'RestElement' ? e.argument.name : e.name) === declarator.id.name
+            );
+          }
+          return declarator.id.name === n.value.left.name;
+        }
+
         if (node.value.declarations.some(d => d.id.name === n.value.left.name)) {
           return true;
         }
       }).size() > 0;
 
-    const hasUpdateMutation = j(node.parent)
+    const hasUpdateMutation = j(scopeNode)
       .find(j.UpdateExpression)
       .filter(n => {
+        if (declarator) {
+          if (declarator.id.type === 'ObjectPattern') {
+            return declarator.id.properties.some(d => d.value.name === n.value.argument.name);
+          } else if (declarator.id.type === 'ArrayPattern') {
+            return declarator.id.elements.some(
+              (e.type === 'RestElement' ? e.argument.name : e.name) === n.value.argument.name
+            ;)
+          }
+
+          return declarator.id.name === n.value.argument.name;
+        }
+
         if (node.value.declarations.some(d => d.id.name === n.value.argument.name)) {
           return true;
         }
       }).size() > 0;
 
-    return hasAssignmentMutation || hasUpdateMutation;
+    return hasAssignmentMutation || hasUpdateMutation
+  }
+
+  const isAccessedInClosure = (node) => {
+
+    return j(node.value.body)
+      .find(j.Identifier)
+      .filter(n => {
+        const declarations = node.value.init ? node.value.init.declarations
+                           : node.value.left ? node.value.left.declarations
+                           : [];
+        if (declarations.some(d => d.id.name === n.value.name)) {
+          let parent = n.parent;
+          while (parent.value !== node.value.body) {
+            parent = parent.parent;
+            if (/Function/.test(parent.value.type)) {
```

Comment:
you can find the types here: https://github.com/benjamn/ast-types/blob/master/def/